dev/TeachingTip/TestUI/TeachingTipPage.xaml.cs

```diff
@@ -0,0 +1,589 @@
+            }
+        }
+        public void OnGetTargetBoundsButtonClicked(object sender, RoutedEventArgs args)
+        {
+            var bounds = this.targetButton.TransformToVisual(null).TransformBounds(new Rect(0.0,
+            0.0,
+            this.targetButton.ActualWidth,
+            this.targetButton.ActualHeight));
+            
+            this.TargetXOffsetTextBlock.Text = bounds.X.ToString();
+            this.TargetYOffsetTextBlock.Text = bounds.Y.ToString();
+            this.TargetWidthTextBlock.Text = bounds.Width.ToString();
+            this.TargetHeightTextBlock.Text = bounds.Height.ToString();
+        }
+
+        public void OnSetScrollViewerOffsetButtonClicked(object sender, RoutedEventArgs args)
+        {
+            this.ContentScrollViewer.ChangeView(0, double.Parse(this.ScrollViewerOffsetTextBox.Text), 1);
+        }
+
+        public void OnBringTargetIntoViewButtonClicked(object sender, RoutedEventArgs args)
+        {
+            this.targetButton.StartBringIntoView(new BringIntoViewOptions());
+        }
+
+        public void OnUseTestWindowBoundsCheckBoxChecked(object sender, RoutedEventArgs args)
+        {
+            Rect windowRect = new Rect(double.Parse(this.TestWindowBoundsXTextBox.Text),
+                                       double.Parse(this.TestWindowBoundsYTextBox.Text),
+                                       double.Parse(this.TestWindowBoundsWidthTextBox.Text),
+                                       double.Parse(this.TestWindowBoundsHeightTextBox.Text));
+            TeachingTipTestHooks.SetUseTestWindowBounds(this.TeachingTip, true);
+            TeachingTipTestHooks.SetTestWindowBounds(this.TeachingTip, windowRect);
+            if(testWindowBounds == null)
+            {
+                testWindowBounds = new Popup();
+                testWindowBounds.IsHitTestVisible = false;
+            }
+            Grid windowBounds = new Grid();
+            windowBounds.Width = windowRect.Width;
+            windowBounds.Height = windowRect.Height;
+            windowBounds.Background = new SolidColorBrush(Colors.Transparent);
+            windowBounds.BorderBrush = new SolidColorBrush(Colors.Red);
+            windowBounds.BorderThickness = new Thickness(1.0);
+            testWindowBounds.Child = windowBounds;
+            testWindowBounds.HorizontalOffset = windowRect.X;
+            testWindowBounds.VerticalOffset = windowRect.Y;
+            testWindowBounds.IsOpen = true;
+        }
+
+        public void OnUseTestWindowBoundsCheckBoxUnchecked(object sender, RoutedEventArgs args)
+        {
+            TeachingTipTestHooks.SetUseTestWindowBounds(this.TeachingTip, false);
+            testWindowBounds.IsOpen = false;
+        }
+
+        public void OnSetPlacementButtonClicked(object sender, RoutedEventArgs args)
+        {
+            if (this.PlacementComboBox.SelectedIndex == 0)
```

Comment:
Avoid comparing against magic numbers.

There are a couple of alternative options here:
1. Compare against the string content of the ComboBoxItem
2. Add x:Names to the ComboBoxItems and do an instance compare.
3. Add enum values directly to the ComboBox, and compare against the enum.

Reply:
Same feedback for elsewhere in this file.

---
In reply to: [252904596](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r252904596) [](ancestors = 252904596)